encoding/types.go

```diff
@@ -0,0 +1,78 @@
+package encoding
+
+import (
+	"io"
+	"reflect"
+
+	"github.com/docker/libchan"
+)
+
+// ChanProducer represents an object which is able to create new
```

Comment:
I might call this a `ChannelFactory` to differentiate it from the notion of a "producer", queuing terminology.